### PR TITLE
gha: Run kubernetes Conformance and SIG-network tests

### PR DESCRIPTION
**Pull request by aojea**
- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

The kubernetes e2e sig-network tests are created to standardize the expected behaviors for the different implementations.

Running them ensure that all the ecosystem provide the same experience for users and contribute to reduce toil to implementations, knowing that the behavior is conformance with the defined by Kubernetes.

### CODE REVIEW

File: .github/workflows/conformance-k8s-kind.yaml
Comment (squeed) [inline .github/workflows/conformance-k8s-kind.yaml]: This should have the `--wait` flag (although I bet kubetest also waits for all the nodes to be Ready). Still good practice.

**Review by squeed — CHANGES_REQUESTED** [review]
one very minor change, then this looks good to me.

**Comment by squeed** [comment]
`--skip="Feature|Federation|PerformanceDNS|DualStack|Disruptive|Serial|KubeProxy|LoadBalancer|GCE|Netpol|NetworkPolicy|rejected|externalTrafficPolicy|HostPort|ExternalIP|kube-proxy" `

This has a lot of skips. Should we be creating issues to un-Skip any of them?

It seems like, at least, NetworkPolicy should be enabled.

**Comment by aojea** [comment]
> `--skip="Feature|Federation|PerformanceDNS|DualStack|Disruptive|Serial|KubeProxy|LoadBalancer|GCE|Netpol|NetworkPolicy|rejected|externalTrafficPolicy|HostPort|ExternalIP|kube-proxy" `
> 
> This has a lot of skips. Should we be creating issues to un-Skip any of them?
> 
> It seems like, at least, NetworkPolicy should be enabled.

remember how did we have it in ovn-kubernetes?

I wanted to do something similiar, to read them from a file adding the issues , so we can work towards remove the list of skips to zero ... can I add this file in the repo? where would be the best place?

**Comment by squeed** [comment]
> I wanted to do something similiar, to read them from a file adding the issues , so we can work towards remove the list of skips to zero ... can I add this file in the repo? where would be the best place?

I think this is fine for now; as we knock the skips down, we can consider moving it. I suspect that the skips will vary with certain configurations (e.g. dualstack, kpr, aws-cni chaining, etc).

**Comment by aojea** [comment]
> I think this is fine for now; as we knock the skips down, we can consider moving it. I suspect that the skips will vary with certain configurations (e.g. dualstack, kpr, aws-cni chaining, etc).

do you want to add a matrix ipv4, ipv6 and dual-stack?
that is very simple to add here

File: .github/workflows/conformance-k8s-kind.yaml
Comment (sayboras) [inline .github/workflows/conformance-k8s-kind.yaml]: just a question, can you share the repo for this, so that we can configure renovate bot to upgrade for the new k8s release?

**Review by sayboras — APPROVED** [review]
Thanks and LGTM 💯 🎖️

File: .github/workflows/conformance-k8s-kind.yaml
Comment (aojea) [reply .github/workflows/conformance-k8s-kind.yaml]: can you elaborate?

**Comment by aojea** [comment]
Last push removes the external dependency on the kind config and add a matrix to test ipv4, ipv6 and dual-stack.

I have this action setup running for years in several projects with success, we can always remove permutation from the matrix slice if we don't want to proceed with all of them

File: .github/workflows/conformance-k8s-kind.yaml
Comment (squeed) [reply .github/workflows/conformance-k8s-kind.yaml]: Ideally we automatically pick up k8s patch releases via renovate. "Minor" releases would still need to be automatic.

Given that we bump k8s versions manually on other places, it's not critical.

Comment (sayboras) [reply .github/workflows/conformance-k8s-kind.yaml]: which repo is the conformance test residing ? Is it part of k8s main repo ? Basically, I would love to configure datasource for renovate bot to update if there is new release?

File: .github/workflows/conformance-k8s-kind.yaml
Comment (aojea) [reply .github/workflows/conformance-k8s-kind.yaml]: check line 148, it sets it to false, for some reasons?

File: .github/workflows/conformance-k8s-kind.yaml
Comment (aojea) [inline .github/workflows/conformance-k8s-kind.yaml]: @squeed I copied this from the other jobs

File: .github/workflows/conformance-k8s-kind.yaml
Comment (aojea) [reply .github/workflows/conformance-k8s-kind.yaml]: this is stored on kubernetes release buckets, it is created every time a new release is cut

But I changed it to match the kind node images, so the e2e and the kubernetes version match, there is no guarantee that e2e tests work against different versions, but is not common we break that though

Comment (sayboras) [reply .github/workflows/conformance-k8s-kind.yaml]: thanks both :100:

**Comment by squeed** [comment]
3 failures, I assume #24174 and #24202 will take care of this
```
 Summarizing 3 Failures:
  [FAIL] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols
  test/e2e/network/service.go:1421
  [FAIL] [sig-network] Services [It] should be able to connect to terminating and unready endpoints if PublishNotReadyAddresses is true
  test/e2e/network/util.go:177
  [FAIL] [sig-network] Services [It] should create endpoints for unready pods
```

**Review by squeed — APPROVED** [review]

**Review by aditighag — COMMENTED** [review]
Looks like some of the tests overlap with the existing [k8s conformance test suite](https://github.com/cilium/cilium/blob/master/test/kubernetes-test.sh) in cilium? Probably good to bring this up in the community meeting. /cc @aanm

**Comment by aojea** [comment]
impressive runnerrs, it takes only 11 mins to run the whole suite of conformance and sig-network

ipv4 and dual-stack fails the expected
```
Summarizing 3 Failures:
  [FAIL] [sig-network] Services [It] should create endpoints for unready pods
  test/e2e/network/service.go:1903
  [FAIL] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols
  test/e2e/network/service.go:1421
  [FAIL] [sig-network] Services [It] should be able to connect to terminating and unready endpoints if PublishNotReadyAddresses is true
  test/e2e/network/util.go:177

Ran 371 of 7069 Specs in 650.984 seconds
FAIL! -- 368 Passed | 3 Failed | 0 Pending | 6698 Skipped


Ginkgo ran 1 suite in 11m0.008641658s
```
```
Summarizing 3 Failures:
  [FAIL] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols
  test/e2e/network/service.go:1421
  [FAIL] [sig-network] Services [It] should be able to connect to terminating and unready endpoints if PublishNotReadyAddresses is true
  test/e2e/network/util.go:177
  [FAIL] [sig-network] Services [It] should create endpoints for unready pods
  test/e2e/network/service.go:1903

Ran 372 of 7069 Specs in 740.912 seconds
FAIL! -- 369 Passed | 3 Failed | 0 Pending | 6697 Skipped
```

ipv6 fails to install cilium  https://github.com/cilium/cilium/actions/runs/4353373283/jobs/7607811754

@squeed do you want to take a look to IPv6 or I just skip it?

**Comment by aojea** [comment]
> Looks like some of the tests overlap with the existing [k8s conformance test suite](https://github.com/cilium/cilium/blob/master/test/kubernetes-test.sh) in cilium? Probably good to bring this up in the community meeting. /cc @aanm

that seems to use vagrant, kind is much better suited for running k8s e2e tests since is one of the kubernetes default tool for testing the project itself and is tuned for performance, see the results

> Ran 371 of 7069 Specs in 650.984 seconds

https://github.com/cilium/cilium/pull/24209#issuecomment-1458408305

 (and I'm also one of the maintainers of KIND 😄 )

**Comment by joestringer** [comment]
> > Looks like some of the tests overlap with the existing [k8s conformance test suite](https://github.com/cilium/cilium/blob/master/test/kubernetes-test.sh) in cilium? Probably good to bring this up in the community meeting. /cc @aanm

> that seems to use vagrant, kind is much better suited for running k8s e2e tests since is one of the kubernetes default tool for testing the project itself and is tuned for performance, see the results

I agree - the current mechanism was built some years ago based on the infra we used at the time. It currently requires a special trigger phrase and relies on some additional custom Jenkins configuration. I like the idea of moving this to GHA instead.

The only thing I would suggest is to check the end of that script for which tests are being skipped + comments and maybe copy any relevant explanations+links here.

At a glance it also looks like that job tests Cilium with kube-proxy, whereas this workflow tests without kube-proxy. cf. https://github.com/cilium/cilium/issues/24199

**Comment by joestringer** [comment]
Do we need to run sig-storage etc.? Some of the reason behind the limited set of tests in the current script is that there's a bunch of stuff that is unrelated to Cilium's functionality.

**Comment by aojea** [comment]
> Do we need to run sig-storage etc.? Some of the reason behind the limited set of tests in the current script is that there's a bunch of stuff that is unrelated to Cilium's functionality.

this regex that I added is specific for networking plugins,

**Comment by sayboras** [comment]
> ipv6 fails to install cilium https://github.com/cilium/cilium/actions/runs/4353373283/jobs/7607811754

You can take a look at the existing ipv6 (e.g. test-smoke-ipv6) jobs and kind-config-ipv6.yaml for reference. The last time I worked on this, I need the bump CIDR range as the default values might be too small. IMO, we can add ipv6 test once in subsequent PR.

Relates:
https://github.com/cilium/cilium/blob/master/.github/workflows/tests-smoke-ipv6.yaml#L129-L133
https://github.com/cilium/cilium/commit/0ac8c4cd3cdafc42b4f122e05d2788a159fb6a57

**Comment by aojea** [comment]
Since my inmidiate priority is to have a Kubernetes job running I prefer to focus on having just ipv4 and monitor the job for some weeks so you can be confident is stable, I tested it locally right now and with my two PRs that got merged there are no failing tests

```

Ran 371 of 7084 Specs in 682.514 seconds
SUCCESS! -- 371 Passed | 0 Failed | 0 Pending | 6713 Skipped
```

**Comment by aojea** [comment]
and this is green now 👍

File: .github/workflows/conformance-k8s-kind.yaml
Comment (sayboras) [inline .github/workflows/conformance-k8s-kind.yaml:93]: this step seems only required once we enable ipv6 testing.

**Review by sayboras — APPROVED** [review]
One small comment, the rest looks good to me. Thanks.

File: .github/workflows/conformance-k8s-kind.yaml
Comment (aojea) [reply .github/workflows/conformance-k8s-kind.yaml:93]: yeah, check that the step is skipped in that case
```
        if: ${{ matrix.ipFamily == 'ipv6' }}
```

Comment (sayboras) [reply .github/workflows/conformance-k8s-kind.yaml:93]: yup, I saw that, the comment is mainly for sanity purposes (i.e. we could just remove this step outright).

I am happy to accept this patch as what we have now.

Comment (aojea) [reply .github/workflows/conformance-k8s-kind.yaml:93]: thanks

**Comment by sayboras** [comment]
Full CI is not required as the changes in this PR are to add a new GHA, which is successfully run.

**Comment by christarazi** [comment]
Removing myself in favor of @nbusseneau

**Comment by aojea** [comment]
just for clarification, both the kubernetes and e2e tests are pinned to a specific version, there is no possibility a test that was passing start to fail if is not related to a change in the cilium code or an environment problem (but this runners seems very powerful )

**Comment by squeed** [comment]
@aditighag @brlbil @nbusseneau would you mind reviewing this PR?

**Review by brlbil — APPROVED** [review]

**Review by aditighag — APPROVED** [review]

**Comment by aditighag** [comment]
I think this is good to go, but let's wait for @nbusseneau's review.

**Comment by aditighag** [comment]
It's been a few days since the pending reviews were requested, so let's get the PR merged. @nbusseneau Please follow up with @aojea if you have any follow-ups in mind.

**Comment by nbusseneau** [comment]
Sorry, I missed the pings last week. Had a review and LGTM, great job, thanks!

**Comment by aojea** [comment]
Thanks, I'll be available for any issue related with this job

**Pull request merged**
